### DISCUSSION TITLE
TcUnit roadmap

### DISCUSSION THREAD

**Discussion by Roald87** [tcunit/TcUnit#139] [Ideas]
Hi Jakob,

What is your plan with TcUnit in the future? Do you have a roadmap and would you like to share it? Personally I think it would be great to reduce the boilerplate as we discussed elsewhere. 

Roald.

**Comment by sagatowski** [comment]
Hi @Roald87!
The roadmap for the next version is this:
https://github.com/tcunit/TcUnit/milestone/4

For version 1.2 it's only foreseen to add the functionality to run test suites and individual tests in sequence.

To reduce boilerplate code I think this lies farther away as it would require changes in the API.

**Reply by Roald87** [reply]
> To reduce boilerplate code I think this lies farther away as it would require changes in the API.

I'm afraid so yes. Are you considering this? Or would you rather keep the current API and live with its limitations?

**Reply by sagatowski** [reply]
When you referred to "elsewhere", do you mean this? https://github.com/tcunit/TcUnit/discussions/63

1. Changing the API would mean we would break TcUnit usage for all the current users.
2. Introducing the concept in #63 above (if there exists one), would it still be compatible with 4020 and 4022?
3. How in practice would it work? I think the discussion came to a halt. Many ideas, but not how to actually solve it.
4. Also, compared to basically any other unit testing framework, I don't think there is more "boilerplate" code in TcUnit. The only thing you really need is to create a test-suite and call TEST() and TEST_FINISHED().

At one point or another, we will most likely have a version 2.0 of TcUnit with new API and everything, although I'm not sure of how this would look like at this point. When doing a 2.0, we most likely won't need to think about all 4020-users, which was the most common version of TwinCAT at the point of release of TcUnit 1.

**Reply by Roald87** [reply]
Yeah mostly #63, but also some discussion in #107. 

> How in practice would it work? I think the discussion came to a halt. Many ideas, but not how to actually solve it.

Yeah I think there is no solution for this at the moment. I think the conclusion was that methods do not provide much flexibility and there is no better way.

> Also, compared to basically any other unit testing framework, I don't think there is more "boilerplate" code in TcUnit. The only thing you really need is to create a test-suite and call TEST() and TEST_FINISHED().

Well, you have to type the name of the test three times (1)when you create it, 2) where you call it and 3) in `TEST()`) and when you rename the test method, the test name doesn't automatically change. In most unit test libraries, you define a test method once and it gets discovered (e.g. by pre-fixing it with `test_`  in Python) and afterwards it gets executed automatically.

**Reply by sagatowski** [reply]
You don't *need* to create separate test-methods for each test. It's just a way to structure the tests. You can run all your tests in the body of the test-suite if you want too, even though that's not preferred as it's not very structured. So although in principle you don't need the test-methods, in practice it makes sense to have them (collecting all local/instance variables in one place for the test etc).

I agree that using a test-method as identifier for a test is better and was what I wanted to have already in version 1.0 of TcUnit. It has many other advantages too, such that you're guranteed to have a unique test-name in the test-suite. But (and that's a big BUT) as it doesn't seem to be technically possible with TwinCAT it doesn't make any sense for me to put it into a roadmap :expressionless:

**Reply by Roald87** [reply]
> You don't need to create separate test-methods for each test.

I saw you mention it once indeed, but I've never done it. Things need to be organized! ;D

> it doesn't make any sense for me to put it into a roadmap

That is true. I was just curious about future development. Maybe you came up with a sudden insight about how to solve this issue, but I guess it is not possible at the moment :(. Anyway I think in the current state TcUnit is already very good, especially after v1.2 with the sequential tests.

**Reply by sagatowski** [reply]
I've been thinking a little more about this, and maybe we should add all good ideas that we can't have in TcUnit 1.x into a label for the next version of TcUnit? (2.0). It's not a road-map, but at least we won't forget any good ideas for the next release.
